pep-9999.rst

```diff
@@ -0,0 +1,294 @@
+
+  1           0 RESUME                   0
+
+  2           2 LOAD_FAST                0 (lst)
+              4 GET_ITER
+              6 LOAD_FAST_AND_CLEAR      1 (x)
+              8 SWAP                     2
+             10 BUILD_LIST               0
+             12 SWAP                     2
+        >>   14 FOR_ITER                 4 (to 26)
+             18 STORE_FAST               1 (x)
+             20 LOAD_FAST                1 (x)
+             22 LIST_APPEND              2
+             24 JUMP_BACKWARD            6 (to 14)
+        >>   26 END_FOR
+             28 SWAP                     2
+             30 STORE_FAST               1 (x)
+             32 RETURN_VALUE
+
+There is no longer a separate code object, nor creation of a single-use function
+object, nor any need to create and destroy a Python frame.
+
+Isolation of the ``x`` iteration variable is achieved by the combination of the
+new ``LOAD_FAST_AND_CLEAR`` opcode at offset ``6``, which saves any outer value
+of ``x`` on the stack before running the comprehension, and ``30 STORE_FAST``,
+which restores the outer value of ``x`` (if any) after running the
+comprehension.
+
+If the comprehension accesses variables from the outer scope, inlining avoids
+the need to place these variables in a cell, allowing the comprehension (and all
+other code in the outer function) to access them as normal fast locals instead.
+This provides further performance gains.
+
+This PEP only proposes to inline comprehensions occurring inside functions,
+where fast-locals (``LOAD_FAST/STORE_FAST``) are used. Module-level
+comprehensions will continue to create and call a function. Module-level
+comprehensions are generally called only once (when the module is imported), so
+optimizing the performance of calling them is unnecessary, and inlining them
+would require separate code paths in the compiler to handle a module global
+namespace dictionary instead of fast-locals.
+
+
+Backwards Compatibility
+=======================
+
+Comprehension inlining will cause the following visible behavior changes. No
+changes in the standard library or test suite were necessary to adapt to these
+changes in the implementation, suggesting the impact in user code is likely to
+be minimal.
+
+Specialized tools depending on undocumented details of compiler bytecode output
+may of course be affected in ways beyond the below, but these tools already must
+adapt to bytecode changes in each Python version.
+
+locals() includes outer variables
+---------------------------------
+
+Calling ``locals()`` within a comprehension will include all locals of the
+function containing the comprehension. E.g. given the following function::
+
+  def f(lst):
+      return [locals() for x in lst]
+
+Calling ``f([1])`` in current Python will return::
+
+  [{'.0': <list_iterator object at 0x7f8d37170460>, 'x': 1}]
+
+Where ``.0`` is the internal implementation detail of the synthetic sole
+argument to the comprehension "function."
```

Comment:
```suggestion
argument to the comprehension "function".
```

Fix punctuation

Reply:
Interesting. I'd always been taught that periods and commas always go inside quotation marks. Searching it now I find that this is apparently American style, and British style puts them outside. (E.g. https://grammar.yourdictionary.com/grammar/punctuation/does-punctuation-go-inside-quotation-marks.html)

Do we prefer the British style in Python PEPs and docs?

Reply:
I think a lot of American style now also puts punctuation outside of quotation marks, e.g. Wikipedia: https://en.wikipedia.org/wiki/Wikipedia:Manual_of_Style#Punctuation_inside_or_outside . I wouldn't want to impose this style if a PEP author prefers otherwise, but I do think it's more logical to put the punctuation outside the quotes.

Reply:
AIUI it was once recommended to put punctuation inside quotes for the sake of movable type, but computer programmers rapidly settled on the form that has the accuracy needed for this industry. And since movable type isn't very relevant any more and computer programming kinda is, the "inside quotes" pattern is becoming the preferred one in most places.